### PR TITLE
fix: preserve allocation evaluation details for ASSIGNMENT_ERROR

### PR DESCRIPTION
**Pull request by greghuels**
Fix for test updates in https://github.com/Eppo-exp/sdk-test-data/pull/49. Tests will pass once that PR is merged.

### CODE REVIEW

**Review by rasendubi — CHANGES_REQUESTED** [review]

File: src/evaluator.ts [hunk 1]
```diff
@@ -94,6 +96,14 @@ export class Evaluator {
...
+            const { flagEvaluationCode, flagEvaluationDescription } =
+              this.getMatchedEvaluationCodeAndDescription(
+                variation,
+                allocation,
+                split,
+                subjectKey,
+                expectedVariationType,
+              );
             const flagEvaluationDetails = flagEvaluationDetailsBuilder
               .setMatch(
```
[the remark this answers is not in the document]
Comment (rasendubi) [reply src/evaluator.ts:108]: My comment above is probably incorrect as we filter out `expectedVariationType` mismatch earlier.

The JSON parsing still might fail though, throwing an exception and erasing evaluation details.

File: src/client/eppo-client.ts [hunk 2]
```diff
@@ -680,7 +680,7 @@ export default class EppoClient {
         expectedVariationType,
       );
 
-      if (!result.variation) {
+      if (!result.variation || result.flagEvaluationDetails.flagEvaluationCode !== 'MATCH') {
```
[the remark this answers is not in the document]
Comment (greghuels) [reply src/client/eppo-client.ts:683]: @rasendubi the "catch" is basically a catch all for any error. Fixing that would require a pretty big refactor to how this works. Considering the tests for this have been merged, what do you think about pushing this through and handle that edge case in a separate PR? That way we're not blocked on failing tests if an urgent issue comes up.

[the remark this answers is not in the document]
Comment (rasendubi) [reply src/client/eppo-client.ts:683]: I'm fine with that although I don't see that as a huge refactor (if `getAssignmentDetail` receives the builder as a parameter instead of creating its own, it's easy to create the builder outside of try-catch here and finish the builder in the catch branch)

[the remark this answers is not in the document]
Comment (greghuels) [reply src/client/eppo-client.ts:684]: This didn't turn out to be as big of a refactor as I thought, so I just added them to this PR rather than creating a new one.

File: src/client/eppo-client.ts [hunk 3]
```diff
@@ -693,16 +694,23 @@ export default class EppoClient {
...
-        '',
-        [],
-        '',
-        '',
-      ).buildForNoneResult('ASSIGNMENT_ERROR', `Assignment Error: ${error.message}`);
-      return {
-        eppoValue,
-        flagEvaluationDetails,
-      };
+      if (error instanceof FlagEvaluationError && error.flagEvaluationDetails) {
```
Comment (rasendubi) [inline src/client/eppo-client.ts:697]: `EppoValue.valueOf()` on line 692 may throw an error that is not `FlagEvaluationError`. That will loose evaluation details

**Review by rasendubi — APPROVED** [review]

File: src/client/eppo-client.ts (hunk 3 above)
Comment (greghuels) [reply src/client/eppo-client.ts:686]: You're right, it will. I made another update to fix that.

**Review by rasendubi — APPROVED** [review]
👍
